lib/ansible/modules/packaging/os/redhat_repository.py

```diff
@@ -0,0 +1,260 @@
+        if repo_url_re:
+            repo_url = repo_url_re.group(1)
+            continue
+
+        repo_enabled_re = re.match(repo_enabled_re_str, line)
+        if repo_enabled_re:
+            repo_enabled = repo_enabled_re.group(1)
+
+            repo = {"id": repo_id, "name": repo_name, "url": repo_url, "enabled": True if repo_enabled == '1' else False}
+            repo_result.append(repo)
+    return repo_result
+
+
+def repository_list(module, list_parameter):
+    # Get rhsm repository list and format it for the user
+    repo = get_repository_list(module, list_parameter)
+    module.exit_json(changed=False, repositories=repo)
+
+
+def repository_modify(module, state, name):
+    current_repo_list = get_repository_list(module, 'all')
+    name = set(name)
+
+    matched_existing_repo = {}
+    for repoid in name:
+        matched_existing_repo[repoid] = []
+        for repo in current_repo_list:
+            if fnmatch(repo['id'], repoid):
+                matched_existing_repo[repoid].append(repo)
+
+    changed = False
+    results = []
+    rhsm_arguments = ['repos']
+
+    for repoid in matched_existing_repo:
+        if len(matched_existing_repo[repoid]) == 0:
+            results.append("%s is not a valid repository ID" % repoid)
+            module.fail_json(results=results, msg="%s is not a valid repository ID" % repoid)
+        for repo in matched_existing_repo[repoid]:
+            if state == 'disabled':
+                if repo['enabled']:
+                    changed = True
+                if module.check_mode:
+                    results.append("Repository '%s' is disabled for this system" % repo['id'])
+                else:
+                    rhsm_arguments += ['--disable', repoid]
+            elif state == 'enabled':
+                if not repo['enabled']:
+                    changed = True
+                if module.check_mode:
+                    results.append("Repository '%s' is enabled for this system" % repo['id'])
+                else:
+                    rhsm_arguments += ['--enable', repoid]
+
+    if module.check_mode:
+        module.exit_json(results=results, changed=changed)
+    rc, out, err = run_subscription_manager(module, rhsm_arguments)
+    results = out.split('\n')
+    module.exit_json(results=results)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            list=dict(choices=['all', 'enabled', 'disabled'],
+                      required=False),
+            name=dict(type="list",
+                      required=False),
+            state=dict(choices=['enabled', 'disabled'],
+                       default=None,
+                       required=False),
+        ),
+        supports_check_mode=True
+    )
+    list = module.params['list']
+    name = module.params['name']
+    state = module.params['state']
+
+    if state is None and list is None:
+        err_msg = "Use list or state option, one of them is required"
```

Comment:
What about to set `err_msg = None` just above the first `if` condition, change all following `if`s to `elif` and then remove the all `module.fail_json(msg=err_msg)` and put one extra `if` at the end to fail when `err_msg is not None`? Something like this:

```
err_msg = None

if state is None and list is None:
    err_msg = "Use list or state option, one of them is required"
elif state is not None and list is not None:
    err_msg = "Both list and state option are defined, use only one of them in the task"
elif list is not None and name is not None:
    err_msg = "When the option list is defined, the option name must not be defined"
elif state in ['enabled', 'disabled'] and name is None:
    err_msg = "When state is enabled or disabled, the option name is required"

if err_msg is not None:
    module.fail_json(msg=err_msg)
```

That should save couple of lines.